zookeeper-server/src/main/java/org/apache/zookeeper/common/X509Util.java

```diff
@@ -548,32 +546,30 @@ public static X509TrustManager createTrustManager(
         boolean ocspEnabled,
         final boolean serverHostnameVerificationEnabled,
         final boolean clientHostnameVerificationEnabled,
-        final boolean fipsMode,
-        final SslRevocationEnabled revocationEnabled) throws TrustManagerException {
+        final boolean fipsMode) throws TrustManagerException {
         if (trustStorePassword == null) {
             trustStorePassword = "";
         }
         try {
             KeyStore ts = loadTrustStore(trustStoreLocation, trustStorePassword, trustStoreTypeProp);
             PKIXBuilderParameters pbParams = new PKIXBuilderParameters(ts, new X509CertSelector());
             if (crlEnabled || ocspEnabled) {
-                if (revocationEnabled == SslRevocationEnabled.LEGACY) {
-                    pbParams.setRevocationEnabled(true);
+                // See [RevocationChecker][1] for details. Basically, we are mimicking legacy path as
+                // that is the path we are routing before(i.e. no explicit `PKIXRevocationChecker`).
+                //
+                // [1]: https://github.com/openjdk/jdk/blob/jdk-11%2B28/src/java.base/share/classes/sun/security/provider/certpath/RevocationChecker.java#L98
+                Set<PKIXRevocationChecker.Option> options = new HashSet<>();
+                if (!ocspEnabled) {
+                    options.add(PKIXRevocationChecker.Option.NO_FALLBACK);
+                    options.add(PKIXRevocationChecker.Option.PREFER_CRLS);
                 }
-                System.setProperty("com.sun.net.ssl.checkRevocation", "true");
-                System.setProperty("com.sun.security.enableCRLDP", "true");
-                if (ocspEnabled) {
-                    Security.setProperty("ocsp.enable", "true");
+                if (Boolean.parseBoolean(Security.getProperty("com.sun.security.onlyCheckRevocationOfEECert")))  {
+                    options.add(PKIXRevocationChecker.Option.ONLY_END_ENTITY);
                 }
+                PKIXRevocationChecker revocationChecker = (PKIXRevocationChecker) CertPathValidator.getInstance("PKIX").getRevocationChecker();
+                revocationChecker.setOptions(options);
+                pbParams.addCertPathChecker(revocationChecker);
             } else {
-                if (revocationEnabled == SslRevocationEnabled.LEGACY) {
-                    pbParams.setRevocationEnabled(false);
-                }
-            }
-
-            if (revocationEnabled == SslRevocationEnabled.TRUE) {
-                pbParams.setRevocationEnabled(true);
-            } else if (revocationEnabled == SslRevocationEnabled.FALSE) {
                 pbParams.setRevocationEnabled(false);
```

Comment:
This is still a problem.

The patch as-is fixes the original issus (clobbering JVM globals), but it still does not allow setting revocation just by the JVM properties.

We can split that problem to a different ticket, but we still need a config option that lets that JVM options work transparently.

Reply:
You are right.

I'd still prefer that there would be a code path that left the setting as default.

Hypothetically, the security provider (or future java versions) may have some alternative methods to specify revocation, which this code is not aware of.

Reply:
> I'd still prefer that there would be a code path that left the setting as default.

Which "setting" do you mean ? This pr default `zookeeper.ssl.crl` to jvm property "com.sun.net.ssl.checkRevocation" and `zookeeper.ssl.ocsp` to security "ocsp.enable".

> the security provider (or future java versions) may have some alternative methods to specify revocation

This is the value of provider and public api, right ? We code to whatever jdk export but not any concrete providers.

See also:
* https://github.com/bcgit/bc-java/blob/main/prov/src/main/java/org/bouncycastle/jce/provider/ProvRevocationChecker.java#L126